pex/interpreter.py

```diff
@@ -1444,10 +1461,56 @@ def spawn_python_job(
         # need to set `__PEX_UNVENDORED__`. See: vendor.__main__.ImportRewriter._modify_import.
         subprocess_env["__PEX_UNVENDORED__"] = "1"
 
-        pythonpath.extend(third_party.expose(expose))
+        pythonpath.extend(third_party.expose(expose, interpreter=interpreter))
 
     interpreter = interpreter or PythonInterpreter.get()
     cmd, process = interpreter.open_process(
         args=args, pythonpath=pythonpath, env=subprocess_env, **subprocess_kwargs
     )
     return Job(command=cmd, process=process)
+
+
+# See the "Test results from various systems" table here:
+#  https://www.in-ulm.de/~mascheck/various/shebang/#length
+MAX_SHEBANG_LENGTH = 512 if sys.platform == "darwin" else 128
+
+
+def create_shebang(
+    python_exe,  # type: str
+    python_args=None,  # type: Optional[str]
+    max_shebang_length=MAX_SHEBANG_LENGTH,  # type: int
+):
+    # type: (...) -> str
+    """Return the contents of an appropriate shebang for the given Python interpreter and args.
+
+    The shebang will include the leading `#!` but will not include a trailing new line character.
+    """
+    python = "{exe} {args}".format(exe=python_exe, args=python_args) if python_args else python_exe
+    shebang = "#!{python}".format(python=python)
+
+    # N.B.: We add 1 to be conservative and account for the EOL character.
+    if len(shebang) + 1 <= max_shebang_length:
+        return shebang
+
+    # This trick relies on /bin/sh being ubiquitous and the concordance of:
```

Comment:
Pip does something like this, but we did not benefit from it since we re-write script shebangs to `#!python` when installing wheels at build time in chroots. The venv 3rdparty dep population code has been changed below though to now use this when it reifies shebangs; so that usage amounts to a bug fix. I could find no Pants or Pex issues though recording someone as having hit a PEX venv too-long shebang though.

Reply:
There was a (what I think could be) related thread in Pants slack #general recently: https://pantsbuild.slack.com/archives/C046T6T9U/p1700074427567259
([linen link](https://chat.pantsbuild.org/t/16066244/we-had-an-unfortunate-regression-in-one-of-our-ci-systems-wh#24929395-3726-4a7a-af1c-347a19798ccb))

Reply:
It looks like that issue would be solved by this, but it could be solved otherwise. That's scie pants using a console script as argv0 (`["pants", ...]`) instead of using Python, a real binary, as argv0 (`[<scie PBS Python>, "pants", ...]). So you all can self serve here.

Reply:
If 2 args vs 1 is a problem, then you can also just emit a small /bin/sh script to use instead as the 1 arg. You'll grow the same system dependence / fragility by relying on a non-hermetic /bin/sh just as you would if you punt to a Pex fix.